src/WebJobs.Extensions.DurableTask.Analyzers/SyntaxNodeUtils.cs

```diff
@@ -236,5 +255,89 @@ private static bool IsSpecifiedAttribute(AttributeSyntax attribute, string attri
 
             return false;
         }
+
+        internal static string GetQualifiedTypeName(ITypeSymbol typeInfo)
+        {
+            if (typeInfo != null)
+            {
+                if (typeInfo is INamedTypeSymbol)
+                {
+                    var tupleUnderlyingType = ((INamedTypeSymbol)typeInfo).TupleUnderlyingType;
+                    if (tupleUnderlyingType != null)
+                    {
+                        return $"System.Tuple<{string.Join(", ", tupleUnderlyingType.TypeArguments.Select(x => x.ToString()))}>";
+                    }
+
+                    return typeInfo.ToString();
+                }
+
+                var arrayString = "";
+                if (typeInfo.Kind.Equals(SymbolKind.ArrayType))
+                {
+                    arrayString = "[]";
+                    typeInfo = ((IArrayTypeSymbol)typeInfo).ElementType;
+                }
+
+                if (!string.IsNullOrEmpty(typeInfo.Name))
+                {
+                    return typeInfo.ContainingNamespace?.ToString() + "." + typeInfo.Name.ToString() + arrayString;
+                }
+            }
+
+            return "Unknown Type";
+        }
+
+        internal static bool InputMatchesOrCompatibleType(ITypeSymbol invocationType, string invocationTypeName, ITypeSymbol functionType, string functionTypeName)
+        {
+            if (invocationType == null || invocationTypeName == null || functionType == null || functionTypeName == null)
+            {
+                return false;
+            }
+
+            return invocationTypeName.Equals(functionTypeName) || AreCompatibleIEnumerableTypes(invocationType, functionType);
+        }
+
+        private static bool AreCompatibleIEnumerableTypes(ITypeSymbol invocationType, ITypeSymbol functionType)
+        {
+            if ((invocationType.Kind.Equals(SymbolKind.ArrayType) || invocationType.Kind.Equals(SymbolKind.NamedType))
+                && (functionType.Kind.Equals(SymbolKind.ArrayType) || functionType.Kind.Equals(SymbolKind.NamedType)))
+                if (UnderlyingTypesMatch(invocationType, functionType))
+                {
+                    return ((invocationType.AllInterfaces.Any(i => i.Name.Equals("IEnumerable")))
+                    && (functionType.AllInterfaces.Any(i => i.Name.Equals("IEnumerable"))));
```

Comment:
UnderlyingTypesMatch checks if I'm looking at an ArrayType or NamedType (like List or Enumerable; by the way, other classes like string implement IEnumerable), then gets and compares the T from both of them. Then I check that they both implement IEnumerable, because any type that implements IEnumerable is serialized to a JSON Array.